keps/sig-node/2033-rootless/README.md

```diff
@@ -0,0 +1,806 @@
+CNI plugins such as Flannel are expected to be used for the containers' namespace.
+
+##### CNI plugins
+
+Flannel (VXLAN) is known to work.
+
+#### Pod SecurityContext
+
+- `runAsUser`: supported, but the number of the UID is limited by `/etc/subuid`.
+- `sysctls`: some sysctl parameters are supported, but some would fail in `EPERM`.
+  Creating Pod manifests with such sysctl parameters would fail.
+  If this behavior is problematic, user should write a Mutating Admission Webhook to remove such sysctl parameters from Pod manifests.
+- seccomp: supported
+- AppArmor: unsupported. Creating Pod manifests with AppArmor profile specified would fail.
+- [Node-level UserNS KEP](https://github.com/kubernetes/enhancements/issues/127): can be supported. This UserNS will be nested inside the Rootless mode's UserNS.
+
+#### Required changes to Kubernetes
+`kubelet` needs to be updated to ignore errors during [`setupKernelTunables()`](https://github.com/kubernetes/kubernetes/blob/v1.19.0-beta.2/pkg/kubelet/cm/container_manager_linux.go#L389-L428),
+because some of sysctl parameters cannot be modified in the user namespace, and modifying them is not mandatory.
+
+> **Note**
+> These sysctl parameters are set by `kubelet` itself and unrelated to `.spec.securityContext.sysctls` in Pod manifests.
+
+Same applies to the `setRLimit()` and the conntrack stuff in `kube-proxy`.
+
+**Patches**
+- [kubelet/cm: ignore sysctl error when running in userns](https://github.com/rootless-containers/usernetes/blob/v20200701.0/src/patches/kubernetes/0001-kubelet-cm-ignore-sysctl-error-when-running-in-usern.patch)
+- [kube-proxy: allow running in userns](https://github.com/rootless-containers/usernetes/blob/v20200701.0/src/patches/kubernetes/0002-kube-proxy-allow-running-in-userns.patch)
+
+### cgroup
+
+[cgroup v2](../20191118-cgroups-v2.md) and systemd are required. cgroup v1 won't be supported due to security concerns.
+
+containerd supports cgroup v2 for rootless since containerd v1.4.
+
+#### Required changes to Kubernetes
+
+`pkg/kubelet/cm` needs to be updated to set `PodConfig.Linux.CgroupParent` to `"user.slice"` so that CRI/OCI can talk to the user instance of systemd.
+A container is executed in a cgroup like "/user.slice/user-1001.slice/user@1001.service/user.slice/cri-containerd-6938564e754ad86269a536ed19172af0eff5f7ff0a41eefc5b433a76f1f38f16.scope".
+
+This feature will be enabled via a boolean field `Rootless` in `KubeletConfiguration`, which will be gated by the `Rootless` feature gate.
+
+**Example kubelet config**
+```yaml
+kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+...
+featureGates:
+  Rootless: true
+rootless: true
+cgroupDriver: systemd
+cgroupsPerQOS: false
+```
+
+**Patches**
+- [kubelet: new feature gate: Rootless](https://github.com/rootless-containers/usernetes/blob/v20200701.0/src/patches/kubernetes/0004-kubelet-new-feature-gate-Rootless.patch)
+
+`cgroupsPerQOS` is currently unsupported.
```

Comment:
why is it not supported?

Reply:
This is because the current patch depends on systemd API, which does not create an arbitrary name of cgroup such as kubepods/{burstable|besteffort}.

But on second thought, it might be possible to just use cgroupfs driver within a custom cgroup root which is created by calling systemd API. In that way, we would be able to create kubepods/{burstable|besteffort}.

I'll try some experiments.

Reply:
Seems to work: https://github.com/rootless-containers/usernetes/pull/200

I'll update this KEP later.

Reply:
Updated KEP to support cgroupsPerQos